Applications/Spire/Source/TimeAndSales/TimeAndSalesTableView.cpp

```diff
@@ -0,0 +1,273 @@
+    indicator_box->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);
+    update_style(*indicator_box, [] (auto& style) {
+      style.get(Any()).
+        set(Visibility::NONE).
+        set(BodyAlign(Qt::AlignHCenter)).
+        set(horizontal_padding(scale_width(8))).
+        set(PaddingBottom(scale_height(20))).
+        set(PaddingTop(scale_height(8)));
+    });
+    return indicator_box;
+  }
+
+  QWidget* item_builder(
+      const std::shared_ptr<TimeAndSalesTableModel>& time_and_sales,
+      const std::shared_ptr<TableModel>& table, int row, int column) {
+    auto column_id = static_cast<TimeAndSalesTableModel::Column>(column);
+    auto cell = [&] () -> QWidget* {
+      if(column_id == TimeAndSalesTableModel::Column::TIME) {
+        auto time = to_text(table->get<ptime>(row, column));
+        return make_label(time.left(time.lastIndexOf('.')));
+      } else if(column_id == TimeAndSalesTableModel::Column::PRICE) {
+        auto money_cell = make_label(to_text(table->get<Money>(row, column)));
+        update_style(*money_cell, apply_table_cell_right_align_style);
+        return money_cell;
+      } else if(column_id == TimeAndSalesTableModel::Column::SIZE) {
+        auto quantity_cell = make_label(
+          to_text(table->get<Quantity>(row, column)).remove(QChar(',')));
+        update_style(*quantity_cell, apply_table_cell_right_align_style);
+        return quantity_cell;
+      } else if(column_id == TimeAndSalesTableModel::Column::MARKET) {
+        return make_label(
+          QString::fromStdString(table->get<std::string>(row, column)));
+      } else if(column_id == TimeAndSalesTableModel::Column::CONDITION) {
+        return make_label(
+          to_text(table->get<TimeAndSale::Condition>(row, column)));
+      }
+      return new QWidget();
+    }();
+    auto indicator = time_and_sales->get_bbo_indicator(row);
+    if(indicator == BboIndicator::UNKNOWN) {
+      match(*cell, UnknownIndicator());
+    } else if(indicator == BboIndicator::ABOVE_ASK) {
+      match(*cell, AboveAskIndicator());
+    } else if(indicator == BboIndicator::AT_ASK) {
+      match(*cell, AtAskIndicator());
+    } else if(indicator == BboIndicator::INSIDE) {
+      match(*cell, InsideIndicator());
+    } else if(indicator == BboIndicator::AT_BID) {
+      match(*cell, AtBidIndicator());
+    } else if(indicator == BboIndicator::BELOW_BID) {
+      match(*cell, BelowBidIndicator());
+    }
+    return cell;
+  }
+}
+
+TableView* Spire::make_time_and_sales_table_view(
+    std::shared_ptr<TimeAndSalesTableModel> table, QWidget* parent) {
+  auto table_view = TableViewBuilder(table).
+    set_header(make_header_model()).
+    set_item_builder(std::bind_front(&item_builder, table)).make();
+  update_style(*table_view, apply_table_view_style);
+  auto& header = table_view->get_header();
+  auto properties = make_header_item_properties();
+  auto table_header_menu = new ContextMenu(header);
```

Comment:
Can you factor this functionality for showing/hiding columns? Have a function defined in TableView that adds this feature, it looks like it could be useful in the future.

Reply:
Do you mean to add a function for toggling column visibility, or to add a menu that provides the showing/hiding functionality to the TableView?

Reply:
I see your purpose. Ignore the above question.